src/util/tmpl/fd_map_perfect.c

```diff
@@ -0,0 +1,349 @@
+the key if you only need a set (containment queries).  To do so,
+#define MAP_PERFECT_HAS_VALUE 0
+And then declare elements in the normal way, e.g.
+    #define MAP_PERFECT_0 44
+adds 44 to the set.
+
+One advanced usage is with multi-element keys (e.g. arrays).  They are
+fully supported, but require #define MAP_PERFECT_COMPLEX_KEY 1.  See
+below for more details. */
+
+#ifndef MAP_PERFECT_NAME
+#error "Define MAP_PERFECT_NAME"
+#endif
+
+/* MAP_PERFECT_LG_TBL_SZ is the base-2 log of the table size.  See the
+   note about MAP_PERFECT_HASH_C below for guidance on how to choose
+   this value.  It must be at least 1. */
+
+#ifndef MAP_PERFECT_LG_TBL_SZ
+#error "Define MAP_PERFECT_LG_TBL_SZ"
+#endif
+
+/* MAP_PERFECT_T is the type of each element in the perfect hash table.
+   There are no requirements on it other than that it must contain the
+   key.  Large structs are okay, since they never get copied, but they
+   might be a pain to const initialize. */
+
+#ifndef MAP_PERFECT_T
+#error "Define MAP_PERFECT_T"
+#endif
+
+/* MAP_PERFECT_KEY is the name of the key field in the struct.  It
+   defaults to key. */
+#ifndef MAP_PERFECT_KEY
+#define MAP_PERFECT_KEY key
+#endif
+
+/* MAP_PERFECT_KEY_T is the type of the query, which should typically be
+   the key type.  When using an array as a key, you may want to make
+   this a const pointer instead of an array though.  For example, if
+   the key is uchar key[32], #define MAP_PERFECT_KEY_T uchar const * is
+   pretty reasonable. */
+
+#ifndef MAP_PERFECT_KEY_T
+#error "Define MAP_PERFECT_KEY_T"
+#endif
+
+/* MAP_PERFECT_COMPLEX_KEY controls whether the key type is a scalar
+   (COMPLEX_KEY==0) or an array (COMPLEX_KEY==1).  If a complex key, the
+   key should be surrounded by parenthesis, e.g.
+     #define MAP_PERFECT_0 (1,1,1), .value=7, .other_value=8
+
+   In this case, when HASH_PP is invoked, each value in the key will be
+   a different argument to the macro.  The hash of the above example
+   will be calculated by expanding MAP_PERFECT_HASH_PP( 1, 1, 1 ), not
+   MAP_PERFECT_HASH_PP( (1,1,1) ). */
+
+#ifndef MAP_PERFECT_COMPLEX_KEY
+#  define MAP_PERFECT_COMPLEX_KEY 0
+#endif
+
+/* MAP_PERFECT_ZERO_KEY must be set to the key of all 0s.  In the
+   non-complex case, the code below probably does the right thing.  The
+   reason for this is a little strange: keys in the table that don't
+   have a value get default initialized, which means set to all 0 bytes.
+   We have to be able to distinguish that case from the case in which
+   the zero key was actually inserted into the table, and we need to be
+   able to do that at preprocessor time.  Especially in the complex key
+   case, this is not easy, so we just require specifying it manually. */
+
+#ifndef MAP_PERFECT_ZERO_KEY
+#  if !MAP_PERFECT_COMPLEX_KEY
+#    define MAP_PERFECT_ZERO_KEY 0
+#  else
+#     error "Define MAP_PERFECT_ZERO_KEY to be the key of all zero bytes"
```

Comment:
No, it has to be usable by the preprocessor. There's an explanation somewhere in there, but basically if nothing you insert hashes to the same thing as the zero key, then you need to insert a dummy element. I couldn't find any other way to prevent false matches for the zero key, since the compiler will initialize any uninitialized elements of the table with all 0s.